Add Default Fallback Marketing Channel Recommendations

ecgan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes issue https://github.com/woocommerce/woocommerce/issues/43032.

In this PR, we add default fallback data for marketing channel recommendations. In relation to that, we also make changes to the code logic (code refactoring) to make use of the Data Source Poller in the code base, so that it is consistent with other features such as payment gateway suggestions, remote free extensions in onboarding wizard flow etc.

**Code references that we use in this PR:**

- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Admin/Features/PaymentGatewaySuggestions
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Internal/Admin/WCPayPromotion
- https://github.com/woocommerce/woocommerce/tree/a9550d8894ea5de1f02fc954ff41db8eac487f73/plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions

**Technical notes on the changes:**

Based on the above code references, I created a new "MarketingRecommendations" feature that replaces the existing "MarketingSpecs". The new feature makes use of the DataSourcePoller class which retrieves recommendation data from woocommerce.com API when the marketing suggestions is turned on, and retrieves recommendation data generated from within WooCommerce plugin (without making request to woocommerce.com API) when the marketing suggestions is turned off. 

With the data source poller, after it retrieves recommendation data from woocommerce.com API, the data will be stored in a transient called `woocommerce_admin_marketing_recommendations_specs`. The transient name is generated by the DataSourcePoller in https://github.com/woocommerce/woocommerce/blob/ccdfb16fc881f56e68d59e9d34810e53841126c7/plugins/woocommerce/src/Admin/DataSourcePoller.php#L68

The old transient `wc_marketing_recommended_plugins` will not be used anymore.

In `DefaultMarketingRecommendations`, most of the code comes from the woocommerce.com API code base. The idea is that when we make changes to the recommendations API in woocommerce.com, we should be able to quickly copy and paste the code change into our WooCommerce plugin code base here with minimal changes.

I also added new icons for the recommended extensions. Some of these icons are not currently being used (e.g. those icons that start with "woocommerce-"), but we want to make it the same as woocommerce.com API and they are being referenced in `DefaultMarketingRecommendations`. In addition, with this in place, in the future we can make default recommendations  show up in Marketing > Coupons page, even when marketing suggestion feature is turned off. 

**Note for next PR:**

Based on the current timeline, this feature shall be released in WooCommerce 8.6.

I will create a separate PR with a separate test instructions to remove the old transient `wc_marketing_recommended_plugins` when the plugin is upgraded from WooCommerce 8.5 to 8.6. 

**Screenshot:**

![image](https://github.com/woocommerce/woocommerce/assets/417342/054858eb-bab7-4ff1-bb27-7b3dd37c023f)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

With marketplace suggestions turned on:

1. Enable marketplace suggestions in WooCommerce Settings > Advanced > Woo.com: `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`
2. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. If there is, delete it. 
3. If there is a transient called `wc_marketing_recommended_plugins`, delete it.
4. Go to Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`
5. We should see marketing channel recommendations. The recommendation data corresponds to the data retrieved from `https://woocommerce.com/wp-json/wccom/marketing-tab/1.3/recommendations.json`.
6. Check that the data is stored in a transient called `woocommerce_admin_marketing_recommendations_specs`. 
7. Check that there is no transient called `wc_marketing_recommended_plugins`.

With marketplace suggestions turned off:

1. Disable marketplace suggestions in WooCommerce Settings > Advanced > Woo.com: `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`
2. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. If there is, delete it.
3. If there is a transient called `wc_marketing_recommended_plugins`, delete it.
4. Go to Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`
5. We should see marketing channel recommendations. The recommendation data corresponds to the data generated in `DefaultMarketingRecommendations.php` file in this PR, with `categories` array containing `'marketing'` and `subcategories` array containing the `'sales-channels'` array. 
    - Note that other subcategories will not be shown in the page because the "Discover more marketing tools" card is not shown when the marketing suggestions feature is turned off. 
7. Check that there isn't a transient called `woocommerce_admin_marketing_recommendations_specs`. 
8. Check that there is no transient called `wc_marketing_recommended_plugins`.

Code base check:

1. Check that the removed code in `MarketingSpecs.php` file are not being used in other places - if they are, they might cause run time errors.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add default fallback marketing channel and tool recommendations.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

ecgan committed:
Add MarketingRecommendations feature.

The code here is based on other feature with default fallbacks and with data source poller, e.g. see payment gateway suggestions.

ecgan committed:
Use MarketingRecommendations feature in API.

ecgan committed:
Add spec_key for MarketingRecommendationsDataSourcePoller.

If we don't specify it, the default is `id`, which is not used in our recommendation data, and this will cause things not to work as expected.

ecgan committed:
Recursively convert object to array.

This is because all the subsequent functions work with arrays, not objects.

ecgan committed:
Make recommended marketing tools and coupons use MarketingRecommendationsInit.

This is so that recommended channels, tools, and coupons make use of the same transient data with data source poller.

ecgan committed:
Add changelog.

ecgan committed:
Fix PHP lint issues.

ecgan committed:
Add doc comment to fix lint error.

ecgan committed:
Fix tests in MarketingRecommendationsTest.

ecgan committed:
Replace recommendations.json file with generated data in DefaultMarketingRecommendations.

This is to have i18n support.

ecgan committed:
Add icons for fallback marketing recommendations.

ecgan committed:
Fix PHP linting issue.

ecgan committed:
Fix linting issue.

ecgan committed:
Merge branch 'trunk' into add/fallback-marketing-recommendations

ecgan committed:
Fix failing tests with PHP 8.0.

ecgan committed:
Remove unused replaced code in MarketingSpecs.

budzanowski commented on plugins/woocommerce/includes/wc-update-functions.php:
```diff
@@ -2460,7 +2460,7 @@ function wc_update_700_remove_download_log_fk() {
  */
 function wc_update_700_remove_recommended_marketing_plugins_transient() {
-	delete_transient( MarketingSpecs::RECOMMENDED_PLUGINS_TRANSIENT );
+	delete_transient( 'wc_marketing_recommended_plugins' );
```
This update function can be deleted because https://github.com/woocommerce/woocommerce/pull/43162 will be introduced. No need to have two update functions doing the same thing in case someone updates from pre-7.0

ecgan replied:
I'm not really sure whether we should do this. 🤔 

Technically, we can delete the function, and things would work as expected.

However, in the way I see it, this `plugins/woocommerce/includes/wc-update-functions.php` file also acts like a "changelog" of database changes. We can look into all the `wc_update_XXX` (where XXX is the 3 digit WooCommerce version number) and know what happened in the database migration during each update. We can even go all the way to version 2.0 from 6 years ago.

If we delete the function, then it feels like we are deleting a record in the "changelog".

To be on the safe side, I'm leaning towards keeping it.

What do you think?

budzanowski replied:
It is redundant this is why I don't like it but I understand the reason behind having it pemanent.

budzanowski commented on plugins/woocommerce/src/Admin/Features/MarketingRecommendations/Init.php:
```diff
@@ -0,0 +1,158 @@
+		$specs = MarketingRecommendationsDataSourcePoller::get_instance()->get_specs_from_data_sources();
+
+		// Fetch specs if they don't yet exist.
+		if ( false === $specs || ! is_array( $specs ) || 0 === count( $specs ) ) {
```
```suggestion
		if ( ! is_array( $specs ) || 0 === count( $specs ) ) {
```
I think that you don't need to check directly for `false` when you are checking for an array with `is_array`. `is_array` will return `false` when `$specs` will be `false` thus making this equivalent.

ecgan replied:
Good point! I also tested this and it works. 👍

ecgan committed:
Remove unneeded false check.

This is because `is_array` would return `false` when `false` is being passed as the argument.

Co-authored-by: Bartosz Budzanowski <bartosz.budzanowski@automattic.com>

budzanowski reviewed (approved):
Looks good!

budzanowski commented:
Please just verify the test before merging.
